### PR TITLE
v3_purp.c: add locking to x509v3_cache_extensions()

### LINKED ISSUE #6121: X509_check_issued doesn't grab locks
I noticed that `X509_check_issued` calls x509v3_cache_extensions without acquiring locks on the individual X509 objects.

It seems that `x509v3_cache_extensions` mutates fields like `ex_flags`. Other usages of `x509v3_cache_extensions` (in `X509_check_purpose`, `X509_check_ca`) acquire the lock before calling.

Given that `X509_check_issued` is an exported function, is there a possibility of a race here?

### PR DESCRIPTION
**Pull request by mspncp**
Fixes #6121

Thanks to Mingtao Yang for reporting this bug.

This is an alternative to pull request #6161.

### CODE REVIEW

**Comment by mspncp** [comment]
Cherry-picks cleanly to 1.1.0 but not to 1.0.2

File: crypto/x509v3/v3_purp.c [hunk 1]
```diff
@@ -354,8 +352,16 @@ static void x509v3_cache_extensions(X509 *x)
     X509_EXTENSION *ex;
 
     int i;
+
     if (x->ex_flags & EXFLAG_SET)
         return;
+
+    CRYPTO_THREAD_write_lock(x->lock);
```
Comment (richsalz) [inline crypto/x509v3/v3_purp.c]: Maybe add a comment that we need to double-check now that we have the lock?

**Comment by mspncp** [comment]
Merged, thanks!

bc624bd95554ea218cbc3fad0d794abdfeb21eb6 (master)
2c2d9dcd25b6a9e9c6522bfd030d6be5d8e1c92a (1.1.0)

**Comment by mspncp** [comment]
*	Maybe add a comment that we need to double-check now that we have the lock?

 

Oh no! I forgot to add the comment before merging! :-/  Please forgive me!

 

Matthias

Von: Matthias St. Pierre <notifications@github.com> 
Gesendet: Donnerstag, 3. Mai 2018 22:38
An: openssl/openssl <openssl@noreply.github.com>
Cc: Dr. Matthias St. Pierre <Matthias.St.Pierre@ncp-e.com>; Your activity <your_activity@noreply.github.com>
Betreff: Re: [openssl/openssl] v3_purp.c: add locking to x509v3_cache_extensions() (#6162)

 

Merged, thanks!

 <https://github.com/openssl/openssl/commit/bc624bd95554ea218cbc3fad0d794abdfeb21eb6> bc624bd (master)
 <https://github.com/openssl/openssl/commit/2c2d9dcd25b6a9e9c6522bfd030d6be5d8e1c92a> 2c2d9dc (1.1.0)

**Pull request closed without merging**